### ISSUE TITLE
One test for some of the metrics in contrib/metrics needs improvement

### ISSUE THREAD

**Issue by sayantan1410** [pytorch/ignite#2517]
## 🐛 Bug description

The tests from :
- [ ]  contrib/metrics/precision_recall_curve.py
- [ ] contrib/metrics/regression/test_median_absolute_error.py
- [ ]  contrib/metrics/regression/test_median_percentage_absolute_error.py
- [ ] contrib/metrics/regression/test_relative_median_absolute_error.py 

 is not correct at the moment but it passes. 

The reason: 
" Each process has the same random state (not sure who is responsible for that, probably torchrun and elastic DDP launch) and thus each process has generated the same y_true and y_preds
 and sklearn precision_recall_curve gives the same result for each process but also the same value for gathered results (in PrecisionRecallCurve) as gathered y_true for example is twice y_true per process (if world size is 2) " - 

Reproducible Code (only for precision_recall_curve.py):
```python
import time
import pytest
import torch
import numpy as np
import ignite.distributed as idist
from ignite.exceptions import NotComputableError
from ignite.metrics import EpochMetric
from ignite.contrib.metrics import PrecisionRecallCurve
from sklearn.metrics import precision_recall_curve
from ignite.engine import Engine


rank = idist.get_rank()
torch.manual_seed(12)
device = idist.device()

torch.set_printoptions(8)


def _test(rank, n_epochs, metric_device):
    metric_device = torch.device(metric_device)
    n_iters = 5
    size = 4
    
    time.sleep(0.1 * rank)
    print(rank, "rng state:", torch.random.get_rng_state().sum())

    y_true = torch.randint(0, 2, (size * n_iters,)).to(device)
    y_preds = torch.randint(0, 2, (size * n_iters,)).to(device)

    def update(engine, i):
        out = (
            y_preds[i * size : (i + 1) * size],
            y_true[i * size : (i + 1) * size],
        )
        return out

    engine = Engine(update)

    prc = PrecisionRecallCurve(device=metric_device)
    prc.attach(engine, "prc")

    data = list(range(n_iters))
    engine.run(data=data, max_epochs=n_epochs)

    assert "prc" in engine.state.metrics

    precision, recall, thresholds = engine.state.metrics["prc"]

    np_y_true = y_true.cpu().numpy().ravel()
    np_y_preds = y_preds.cpu().numpy().ravel()
    
    time.sleep(0.1 * rank)
    print(rank, "np_y_true:", np_y_true)
    print(rank, "np_y_preds:", np_y_preds)

    sk_precision, sk_recall, sk_thresholds = precision_recall_curve(np_y_true, np_y_preds)

    assert precision.shape == sk_precision.shape
    assert recall.shape == sk_recall.shape
    assert thresholds.shape == sk_thresholds.shape

    time.sleep(0.1 * rank)
    print(rank, f"precision: {precision}")
    print(rank, f"recall: {recall}")
    print(rank, f"thresholds: {thresholds}")
    print(rank, f"sk_precision: {sk_precision}")
    print(rank, f"sk_recall: {sk_recall}")
    print(rank, f"sk_thresholds: {sk_thresholds}")
    
    if rank == 0:
        np_y_true = np.stack([np_y_true, np_y_true], axis=0).ravel()
        np_y_preds = np.stack([np_y_preds, np_y_preds], axis=0).ravel()
        sk_precision, sk_recall, sk_thresholds = precision_recall_curve(np_y_true, np_y_preds)
        print(rank, f"sk_precision: {sk_precision}")
        print(rank, f"sk_recall: {sk_recall}")
        print(rank, f"sk_thresholds: {sk_thresholds}")
    
    
with idist.Parallel("gloo") as p:
    p.run(_test, 1, "cpu")
```

Command to run: 
`CUDA_VISIBLE_DEVICES= torchrun --nproc_per_node 2 failing_ddp_test_metrics.py`

Next Step: 
1) First, try to reproduce the error and make sure to be able to make test fail and understand why it is so.
2) Next, try to reuse the code from https://github.com/pytorch/ignite/pull/2511#issuecomment-1066114980  to make generated data different for each rank
3) Now test should pass, print/debug intermediate results to make sure that they are passing not just because we generate some special data such that tests are passing (as now)

**Comment by vfdev-5** [comment]
@sayantan1410 can you extend this issue to other tests that suffer from the same problem

**Comment by vfdev-5** [comment]
Done by @puhuk in his recent PR.
@puhuk can you comment out here and provide the links on related PRs. Thanks!

**Comment by puhuk** [comment]
@sayantan1410 Thank you for the issue. It seems resolved from #2655. It includes generating different data from each rank and ensuring the metric of all gathered data from each rank has same value with reference metric (from sklearn).  :)

**Comment by sayantan1410** [comment]
> @sayantan1410 Thank you for the issue. It seems resolved from #2655. It includes generating different data from each rank and ensuring the metric of all gathered data from each rank has same value with reference metric (from sklearn). :)

Lovely to hear that the issue is finally solved. Thanks and Good Luck !
